src/tir/transforms/storage_flatten.cc

```diff
@@ -50,6 +50,913 @@ using runtime::StorageRank;
+
+    internal_buf_map_.at(key).in_scope = false;
+
+    return stmt;
+  }
+
+  Stmt VisitStmt_(const BufferStoreNode* op) final {
+    Stmt stmt = StmtExprMutator::VisitStmt_(op);
+    op = stmt.as<BufferStoreNode>();
+    ICHECK(op);
+
+    auto it = internal_buf_map_.find(op->buffer);
+    if (it != internal_buf_map_.end()) {
+      const InternalBufferRemap& entry = it->second;
+      ICHECK(entry.in_scope) << "Cannot store to an out-of-scope buffer";
+      ICHECK_EQ(entry.realized_begins.size(), op->indices.size())
+          << "Inconsistent dimensions for buffer " << op->buffer->name;
+
+      Array<PrimExpr> new_indices;
+      for (size_t i = 0; i < entry.realized_begins.size(); i++) {
+        new_indices.push_back(op->indices[i] - entry.realized_begins[i]);
+      }
+
+      BufferStore updated = GetRef<BufferStore>(op);
+      auto write_ptr = updated.CopyOnWrite();
+      write_ptr->indices = new_indices;
+      write_ptr->buffer = entry.remap_to;
+      stmt = updated;
+    }
+
+    return stmt;
+  }
+
+  PrimExpr VisitExpr_(const BufferLoadNode* op) final {
+    PrimExpr expr = StmtExprMutator::VisitExpr_(op);
+    op = expr.as<BufferLoadNode>();
+    ICHECK(op);
+
+    auto it = internal_buf_map_.find(op->buffer);
+    if (it != internal_buf_map_.end()) {
+      const InternalBufferRemap& entry = it->second;
+      ICHECK(entry.in_scope) << "Cannot read from an out-of-scope buffer";
+      ICHECK_EQ(entry.realized_begins.size(), op->indices.size())
+          << "Inconsistent dimensions for buffer " << op->buffer->name;
+
+      Array<PrimExpr> new_indices;
+      for (size_t i = 0; i < entry.realized_begins.size(); i++) {
+        new_indices.push_back(op->indices[i] - entry.realized_begins[i]);
+      }
+
+      BufferLoad updated = GetRef<BufferLoad>(op);
+      auto write_ptr = updated.CopyOnWrite();
+      write_ptr->indices = new_indices;
+      write_ptr->buffer = entry.remap_to;
+      expr = updated;
+    }
+
+    return expr;
+  }
+
+  Stmt VisitStmt_(const AttrStmtNode* op) final {
+    if (op->node->IsInstance<tir::BufferNode>()) {
+      // Visit body before checking internal_buf_map_, because we
+      // don't know if the BufferNode needs to be changed until we
+      // look in the body for a BufferRealizeNode with different
+      // extents.
+      Stmt body = this->VisitStmt(op->body);
+
+      Buffer buffer = Downcast<tir::Buffer>(op->node);
+      auto it = internal_buf_map_.find(buffer);
+      if (it != internal_buf_map_.end()) {
+        buffer = it->second.remap_to;
+        return AttrStmt(it->second.remap_to, op->attr_key, op->value, body);
+      }
+      return AttrStmt(buffer, op->attr_key, op->value, body);
+
+    } else if (op->attr_key == attr::buffer_bind_scope) {
+      return HandleBufferBindScope(op);
+    }
+
+    return StmtExprMutator::VisitStmt_(op);
+  }
+
+ private:
+  Stmt HandleBufferBindScope(const AttrStmtNode* op) {
```

Comment:
Add a comment describing that the buffer bind scope buffer attributes are updated according to the legalized buffer. Similarly add comments in the other passes which have BufferBindScope handler methods.

Reply:
Makes sense, and comments have been added.